File: backend/routes/transport.py
"""
TravelVista — Transport Routes
"""
import time
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import get_db, transport_to_dict

transport_bp = Blueprint('transport', __name__, url_prefix='/api/transport')


def _is_admin(uid):
    conn = get_db()
    row = conn.execute("SELECT role FROM users WHERE id = ?", (uid,)).fetchone()
    conn.close()
    return row and row['role'] == 'admin'


@transport_bp.route('', methods=['GET'])
@jwt_required()
def list_transport():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    rows = conn.execute("SELECT * FROM transport").fetchall()
    conn.close()
    return jsonify([transport_to_dict(r) for r in rows])


@transport_bp.route('', methods=['POST'])
@jwt_required()
def add_transport():
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    trn_id = 'trn-' + str(int(time.time() * 1000))

    conn = get_db()
    conn.execute(
        "INSERT INTO transport (id,name,type,capacity,price_per_day,image) VALUES (?,?,?,?,?,?)",
        (trn_id, data.get('name', ''), data.get('type', 'Car'),
         data.get('capacity', 0), data.get('pricePerDay', 0), data.get('image', ''))
    )
    conn.commit()
    row = conn.execute("SELECT * FROM transport WHERE id = ?", (trn_id,)).fetchone()
    conn.close()
    return jsonify(transport_to_dict(row)), 201


@transport_bp.route('/<trn_id>', methods=['PUT'])
@jwt_required()
def update_transport(trn_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    data = request.get_json() or {}
    conn = get_db()
    conn.execute(
        "UPDATE transport SET name=?, type=?, capacity=?, price_per_day=?, image=? WHERE id=?",
        (data.get('name', ''), data.get('type', 'Car'),
         data.get('capacity', 0), data.get('pricePerDay', 0),
         data.get('image', ''), trn_id)
    )
    conn.commit()
    row = conn.execute("SELECT * FROM transport WHERE id = ?", (trn_id,)).fetchone()
    conn.close()
    if not row:
        return jsonify({'message': 'Vehicle not found'}), 404
    return jsonify(transport_to_dict(row))


@transport_bp.route('/<trn_id>', methods=['DELETE'])
@jwt_required()
def delete_transport(trn_id):
    uid = get_jwt_identity()
    if not _is_admin(uid):
        return jsonify({'message': 'Forbidden'}), 403

    conn = get_db()
    conn.execute("DELETE FROM transport WHERE id = ?", (trn_id,))
    conn.commit()
    conn.close()
    return jsonify({'success': True})

File: api/transport.php
<?php
/**
 * TravelVista — Transport API (Admin)
 * GET                — List transport
 * POST               — Add transport
 * PUT    ?id=X       — Update transport
 * DELETE ?id=X       — Delete transport
 */

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Authorization');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') { http_response_code(200); exit; }

require_once __DIR__ . '/../config/jwt_helper.php';

$method = $_SERVER['REQUEST_METHOD'];
$id = $_GET['id'] ?? '';
if (!$id) {
    $uri = $_SERVER['REQUEST_URI'] ?? '';
    if (preg_match('#/api/transport/(\d+)#', $uri, $m)) { $id = $m[1]; }
}

// ── GET: List transport ──
if ($method === 'GET') {
    require_admin();
    $db = getDB();
    $rows = $db->query("SELECT * FROM transport")->fetchAll();
    json_response(array_map('transport_to_dict', $rows));
}

// ── POST: Add transport ──
if ($method === 'POST') {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $stmt = $db->prepare(
        "INSERT INTO transport (name, type, capacity, price_per_day, image) VALUES (?, ?, ?, ?, ?)"
    );
    $stmt->execute([
        $data['name'] ?? '', $data['type'] ?? 'Car',
        $data['capacity'] ?? 0, $data['pricePerDay'] ?? 0, $data['image'] ?? '',
    ]);

    $trnId = $db->lastInsertId();
    $stmt = $db->prepare("SELECT * FROM transport WHERE id = ?");
    $stmt->execute([$trnId]);
    json_response(transport_to_dict($stmt->fetch()), 201);
}

// ── PUT: Update transport ──
if ($method === 'PUT' && $id) {
    require_admin();
    $data = get_json_body();
    $db = getDB();

    $db->prepare(
        "UPDATE transport SET name=?, type=?, capacity=?, price_per_day=?, image=? WHERE id=?"
    )->execute([
        $data['name'] ?? '', $data['type'] ?? 'Car',
        $data['capacity'] ?? 0, $data['pricePerDay'] ?? 0,
        $data['image'] ?? '', $id,
    ]);

    $stmt = $db->prepare("SELECT * FROM transport WHERE id = ?");
    $stmt->execute([$id]);
    $row = $stmt->fetch();
    if (!$row) json_response(['message' => 'Vehicle not found'], 404);
    json_response(transport_to_dict($row));
}

// ── DELETE: Delete transport ──
if ($method === 'DELETE' && $id) {
    require_admin();
    $db = getDB();
    $db->prepare("DELETE FROM transport WHERE id = ?")->execute([$id]);
    json_response(['success' => true]);
}

json_response(['message' => 'Invalid request'], 400);
?>
